kubernetes/websocket/wsclient.c

```diff
@@ -0,0 +1,331 @@
+#include "wsclient.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <stdbool.h>
+
+#define WSC_ATTACH_STDIN_BUFFER_SIZE 1024
+#define TTY_STDIN_NUMBER 0
+#define TTY_STDOUT_NUMBER 1
+#define WS_PROTOCOL_DELIM "://"
+#define WS_BASE_PATH_DELIM ":"  /* ip:port */
+
+static struct lws_context *g_lws_context;
+static bool g_interrupted;
+
+/*
+ * The retry and backoff policy we want to use for our client connections
+ */
+
+static const uint32_t backoff_ms[] = { 1000, 2000, 3000, 4000, 5000 };
+
+static const lws_retry_bo_t retry = {
+    .retry_ms_table = backoff_ms,
+    .retry_ms_table_count = LWS_ARRAY_SIZE(backoff_ms),
+    .conceal_count = LWS_ARRAY_SIZE(backoff_ms),
+
+    .secs_since_valid_ping = 3, /* force PINGs after secs idle */
+    .secs_since_valid_hangup = 10,  /* hangup after secs idle */
+
+    .jitter_percent = 20,
+};
+
+static int get_server_address_and_port(char **p_ws_addr, int *p_ws_port, const char *base_path)
+{
+    if (!base_path) {
+        return -1;
+    }
+
+    char *p = strstr(base_path, WS_PROTOCOL_DELIM); /* e.g. base_path="https://ip:port" */
+    if (NULL == p) {
+        fprintf(stderr, "%s: %s is not a valid websocket base path.\n", __func__, base_path);
+        return -1;
+    }
+    const char *addr_and_port_string = p + strlen(WS_PROTOCOL_DELIM);   /* addr_and_port_string="ip:port" */
+
+    p = strstr(addr_and_port_string, WS_BASE_PATH_DELIM);
+    if (NULL == p) {
+        fprintf(stderr, "%s: %s is not a valid websocket base path.\n", __func__, addr_and_port_string);
+        return -1;
+    }
+    const char *port_string = p + strlen(WS_BASE_PATH_DELIM);
+
+    int ws_port = atoi(port_string);
+    if (0 == ws_port) {
```

Comment:
I think it's better to use strtol intead of using atoi
atoi is unsafe function.

Reply:
Why is `atoi` unsafe ?

Reply:
afaik, `atoi("some-non-numeric-string")` is undefined in the C spec, so the `strto...` family are preferred.

https://wiki.sei.cmu.edu/confluence/display/c/ERR34-C.+Detect+errors+when+converting+a+string+to+a+number

I don't think it matters too much in this case, but it's a good habit to get into.